File: app/src/main/java/com/bopr/android/smailer/sync/SyncAdapter.java
package com.bopr.android.smailer.sync;

import android.accounts.Account;
import android.content.AbstractThreadedSyncAdapter;
import android.content.ContentProviderClient;
import android.content.Context;
import android.content.SyncResult;
import android.os.Bundle;

import com.bopr.android.smailer.Database;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Handle the transfer of data between a server and an app, using the Android sync adapter framework.
 * <p>
 * Required by synchronization framework.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
class SyncAdapter extends AbstractThreadedSyncAdapter {

    private static final Logger log = LoggerFactory.getLogger("SyncAdapter");

    SyncAdapter(Context context, boolean autoInitialize) {
        super(context, autoInitialize);
    }

    @Override
    public void onPerformSync(Account account, Bundle extras, String authority,
                              ContentProviderClient provider, SyncResult syncResult) {
        log.warn("Synchronizing");

        try(Database database = new Database(getContext())) {
            new Synchronizer(getContext(), account, database).sync();
        } catch (Exception x) {
            log.warn("Synchronization failed ", x);
        }
    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/sync/SyncAdapter.kt
package com.bopr.android.smailer.sync

import android.accounts.Account
import android.content.AbstractThreadedSyncAdapter
import android.content.ContentProviderClient
import android.content.Context
import android.content.SyncResult
import android.os.Bundle
import com.bopr.android.smailer.Database
import org.slf4j.LoggerFactory

/**
 * Handle the transfer of data between a server and an app, using the Android sync adapter framework.
 *
 *
 * Required by synchronization framework.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
/* To debug it (to put breakpoints) remove android:process=":sync" from AndroidManifest */
class SyncAdapter(context: Context, autoInitialize: Boolean) : AbstractThreadedSyncAdapter(context, autoInitialize) {

    private val log = LoggerFactory.getLogger("SyncAdapter")

    override fun onPerformSync(account: Account, extras: Bundle?, authority: String?,
                               provider: ContentProviderClient?, syncResult: SyncResult?) {
        val upload = extras?.getBoolean("expect_upload", false)
        try {
            Database(context).use {
                Synchronizer(context, account, it).sync()
            }
            log.debug("Synchronized")
        } catch (x: Exception) {
            log.warn("Synchronization failed ", x)
        }
    }
}